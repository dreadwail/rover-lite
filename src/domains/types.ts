import { Action, AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { RoverSearchState } from '../state/types';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type GenericThunkAction<T, S extends RoverSearchState = RoverSearchState> = ThunkAction<
  T,
  S,
  {},
  AnyAction
>;
export type PromiseThunkAction<
  T,
  S extends RoverSearchState = RoverSearchState
> = GenericThunkAction<Promise<T>, S>;

export type MapDispatchToProps<
  TDispatchProps,
  TOwnProps = {},
  TState extends RoverSearchState = RoverSearchState,
  A extends Action = Action
> = <E>(dispatch: ThunkDispatch<TState, E, A>, ownProps: TOwnProps) => TDispatchProps;
