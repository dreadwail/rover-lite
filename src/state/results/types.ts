export interface Badge {
  readonly content: string;
  readonly imageUrl: string;
  readonly title: string;
}

export interface Listing {
  readonly thumbnailUrl: string;
  readonly rank: number;
  readonly name: string;
  readonly badges: Badge[];
  readonly tagline: string;
  readonly neighborhood: string;
  readonly city: string;
  readonly state: string;
  readonly zip: string;
  readonly repeatClientCount: number;
  readonly reviewCount: number;
  readonly ratingAverage: number;
  readonly featuredReview: string;
  readonly currencySymbol: string;
  readonly price: number;
  readonly priceUnit: string;
  readonly webUrl: string;
}

export interface Results {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly isStale: boolean;
  readonly listings: Listing[];
}
