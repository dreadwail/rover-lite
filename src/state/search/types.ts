import { SearchType } from '../searchTypes/types';

export type SearchInputs = Record<string, any>;

export interface Search {
  readonly searchType: SearchType;
  readonly latitude: number;
  readonly longitude: number;
  readonly inputs: SearchInputs;
}
