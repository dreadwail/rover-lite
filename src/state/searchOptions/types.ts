export interface OptionCategory {
  readonly categoryId: string;
  readonly rank: number;
  readonly displayName: string;
}

export interface BasicOption {
  readonly type: string;
  readonly required: boolean;
  readonly label: string;
  readonly category: OptionCategory;
  readonly rank: number;
  readonly helpText?: string;
}

export interface DateOption extends BasicOption {
  readonly type: 'date';
}

export interface Choice {
  readonly value: any;
  readonly displayName: string;
}

export interface ChoiceOption extends BasicOption {
  readonly type: 'choice';
  readonly choices: Choice[];
}

export interface MultipleChoiceOption extends BasicOption {
  readonly type: 'multiple choice';
  readonly choices: Choice[];
}

export interface IntegerOption extends BasicOption {
  readonly type: 'integer';
  readonly minValue: number;
  readonly maxValue: number;
}

export interface BooleanOption extends BasicOption {
  readonly type: 'boolean';
}

export type Option =
  | BasicOption
  | DateOption
  | ChoiceOption
  | MultipleChoiceOption
  | IntegerOption
  | BooleanOption;

export const isBooleanOption = (option: Option): option is BooleanOption =>
  option.type === 'boolean';
export const isIntegerOption = (option: Option): option is IntegerOption =>
  option.type === 'integer';
export const isMultipleChoiceOption = (option: Option): option is MultipleChoiceOption =>
  option.type === 'multiple choice';
export const isChoiceOption = (option: Option): option is ChoiceOption => option.type === 'choice';
export const isDateOption = (option: Option): option is DateOption => option.type === 'date';
export const isBasicOption = (option: Option): option is BasicOption =>
  [isBooleanOption, isIntegerOption, isChoiceOption, isMultipleChoiceOption, isDateOption].every(
    check => !check(option)
  );

export type Options = Record<string, Option>;
export type OptionsByCategoryId = Record<string, Options>;

export interface SearchOptions {
  readonly isFetching: boolean;
  readonly hasFailed: boolean;
  readonly options: Options;
}
