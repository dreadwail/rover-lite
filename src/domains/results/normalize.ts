import { Listing } from '../../state/results/types';

export interface RawDerivedData {
  readonly zip_code: string;
}

export interface RawResponsiveness {
  readonly photo_rate?: number;
  readonly photo_rate_desc: string;
  readonly percent?: number;
  readonly percent_desc: string;
  readonly time_short_desc: string;
}

export interface RawImage {
  readonly small: string;
}

export interface RawAttribute {
  readonly text: string;
  readonly key: string;
  readonly icon: string;
}

export interface RawPreferences {
  readonly dogs_preferences_attributes: RawAttribute[];
  readonly accepted_dogs_attributes: RawAttribute[];
  readonly dogs_experience_attributes: RawAttribute[];
  readonly attributes: RawAttribute[];
}

export interface RawProviderProfile {
  readonly hosting_service_preferences: RawPreferences;
  readonly traveling_service_preferences: RawPreferences;
  readonly attributes: RawAttribute[];
  readonly years_of_experience: number;
}

export interface RawBadge {
  readonly icon: string;
  readonly description: string;
  readonly title: string;
}

export interface RawPersonSummary {
  readonly responsiveness: RawResponsiveness;
  readonly first_name: string;
  readonly reviews_count: number;
  readonly profile_photo: RawImage;
  readonly ratings_average: string;
  readonly web_url: string;
  readonly images: RawImage[];
  readonly provider_profile: RawProviderProfile;
  readonly attributes: Record<string, RawAttribute>;
  readonly badges: RawBadge[];
}

export interface RawBadgeData {
  readonly content: string;
  readonly image_url: string;
  readonly show_on_search: boolean;
  readonly title: string;
}

export interface RawReview {
  readonly content: string;
  readonly dog_photo_url: string;
}

export interface RawResult {
  readonly rank: number;
  readonly person_summary: RawPersonSummary;
  readonly reviews_count: number;
  readonly neighborhood: string;
  readonly price_unit: string;
  readonly city: string;
  readonly zip: string;
  readonly title: string;
  readonly state: string;
  readonly contact_url: string;
  readonly person_name: string;
  readonly price: string;
  readonly review_text: string;
  readonly review_dog_photo_url: string;
  readonly badge_data: RawBadgeData[];
  readonly repeat_client_count: number;
  readonly reviews: RawReview[];
  readonly ratings_average: string;
  readonly currency_symbol: string;
}

export interface RawSearchResponse {
  readonly results: RawResult[];
}

const rawResultToListing = (rawResult: RawResult): Listing => ({
  thumbnailUrl: rawResult.person_summary.profile_photo.small,
  rank: rawResult.rank,
  name: rawResult.person_summary.first_name,
  badges: rawResult.badge_data.map(rawBadge => ({
    content: rawBadge.content,
    imageUrl: rawBadge.image_url,
    title: rawBadge.title,
  })),
  tagline: rawResult.title,
  neighborhood: rawResult.neighborhood,
  city: rawResult.city,
  state: rawResult.state,
  zip: rawResult.zip,
  repeatClientCount: rawResult.repeat_client_count,
  reviewCount: rawResult.reviews_count,
  ratingAverage: parseFloat(rawResult.ratings_average),
  featuredReview: rawResult.review_text,
  currencySymbol: rawResult.currency_symbol,
  price: parseFloat(rawResult.price),
  priceUnit: rawResult.price_unit,
  webUrl: rawResult.person_summary.web_url,
});

export const normalize = (rawResponse: RawSearchResponse): Listing[] =>
  rawResponse.results.map(rawResultToListing);
