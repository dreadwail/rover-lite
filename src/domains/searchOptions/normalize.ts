import {
  BasicOption,
  BooleanOption,
  ChoiceOption,
  DateOption,
  IntegerOption,
  MultipleChoiceOption,
  Option,
  Options,
} from '../../state/searchOptions/types';

interface RawBasicOption {
  readonly type: string;
  readonly required: boolean;
  readonly label: string;
  readonly help_text?: string;
  readonly category?: string;
  readonly category_rank?: number;
  readonly category_display_name?: string;
  readonly rank?: number;
}

interface RawDateOption extends RawBasicOption {
  readonly type: 'date';
}

interface RawChoice {
  readonly value: any;
  readonly display_name: string;
}

interface RawChoiceOption extends RawBasicOption {
  readonly type: 'choice';
  readonly choices: RawChoice[];
}

interface RawMultipleChoiceOption extends RawBasicOption {
  readonly type: 'multiple choice';
  readonly choices: RawChoice[];
}

interface RawIntegerOption extends RawBasicOption {
  readonly type: 'integer';
  readonly min_value: number;
  readonly max_value: number;
}

interface RawBooleanOption extends RawBasicOption {
  readonly type: 'boolean';
}

type RawOption =
  | RawBasicOption
  | RawDateOption
  | RawChoiceOption
  | RawMultipleChoiceOption
  | RawIntegerOption
  | RawBooleanOption;

const isRawBooleanOption = (rawOption: RawOption): rawOption is RawBooleanOption =>
  rawOption.type === 'boolean';
const isRawIntegerOption = (rawOption: RawOption): rawOption is RawIntegerOption =>
  rawOption.type === 'integer';
const isRawMultipleChoiceOption = (rawOption: RawOption): rawOption is RawMultipleChoiceOption =>
  rawOption.type === 'multiple choice';
const isRawChoiceOption = (rawOption: RawOption): rawOption is RawChoiceOption =>
  rawOption.type === 'choice';
const isRawDateOption = (rawOption: RawOption): rawOption is RawDateOption =>
  rawOption.type === 'date';

type OptionsResponseActionType = 'GET';

type RawOptions = Record<string, RawOption>;

export interface OptionsResponse {
  readonly name: string;
  readonly description: string;
  readonly renders: string[];
  readonly parses: string[];
  readonly actions: Record<OptionsResponseActionType, RawOptions>;
}

const extractRawCategory = (rawBasicOption: RawBasicOption): Option['category'] => {
  const defaultCategory: Option['category'] = {
    categoryId: 'detail',
    rank: 0,
    displayName: 'Detail',
  };
  if (!rawBasicOption.category) {
    return defaultCategory;
  }
  if (!rawBasicOption.category_display_name) {
    return defaultCategory;
  }
  if (!rawBasicOption.category_rank) {
    return defaultCategory;
  }
  return {
    categoryId: rawBasicOption.category,
    rank: rawBasicOption.category_rank,
    displayName: rawBasicOption.category_display_name,
  };
};

const labelOverrides: Record<string, string> = {
  Minprice: 'Minimum Price',
  Maxprice: 'Maximum Price',
};

const convertOverridableLabel = (rawLabel: string): string => {
  const override = labelOverrides[rawLabel];
  if (override) {
    return override;
  }
  return rawLabel;
};

const rawBasicOptionToOption = (rawBasicOption: RawBasicOption): BasicOption => ({
  type: rawBasicOption.type,
  required: !!rawBasicOption.required,
  label: convertOverridableLabel(rawBasicOption.label),
  helpText: rawBasicOption.help_text,
  rank: rawBasicOption.rank || 0,
  category: extractRawCategory(rawBasicOption),
});

const rawBooleanOptionToOption = (rawBooleanOption: RawBooleanOption): BooleanOption => {
  const option = rawBasicOptionToOption(rawBooleanOption);
  return { ...option, type: 'boolean' };
};

const rawIntegerOptionToOption = (rawIntegerOption: RawIntegerOption): IntegerOption => {
  const option = rawBasicOptionToOption(rawIntegerOption);
  return {
    ...option,
    type: 'integer',
    minValue: rawIntegerOption.min_value,
    maxValue: rawIntegerOption.max_value,
  };
};

const choiceDisplayNameOverrides: Record<string, string> = {
  no_pref: 'No preference',
  no_other_dogs: 'No other dogs',
  other_dogs: 'Other dogs',
  houses: 'Houses',
  apartments: 'Apartments',
};

const convertOverridableChoiceDisplayName = (rawDisplayName: string): string => {
  const override = choiceDisplayNameOverrides[rawDisplayName];
  if (override) {
    return override;
  }
  return rawDisplayName;
};

const rawChoiceOptionToOption = (rawChoiceOption: RawChoiceOption): ChoiceOption => {
  const option = rawBasicOptionToOption(rawChoiceOption);
  return {
    ...option,
    type: 'choice',
    choices: rawChoiceOption.choices.map(choice => ({
      value: choice.value,
      displayName: convertOverridableChoiceDisplayName(choice.display_name),
    })),
  };
};

const rawMultipleChoiceOptionToOption = (
  rawMultipleChoiceOption: RawMultipleChoiceOption
): MultipleChoiceOption => {
  const option = rawBasicOptionToOption(rawMultipleChoiceOption);
  return {
    ...option,
    type: 'multiple choice',
    choices: rawMultipleChoiceOption.choices.map(choice => ({
      value: choice.value,
      displayName: convertOverridableChoiceDisplayName(choice.display_name),
    })),
  };
};

const rawDateOptionToOption = (rawDateOption: RawDateOption): DateOption => {
  const option = rawBasicOptionToOption(rawDateOption);
  return { ...option, type: 'date' };
};

const rawOptionToOption = (rawOption: RawOption): Option => {
  if (isRawBooleanOption(rawOption)) {
    return rawBooleanOptionToOption(rawOption);
  }
  if (isRawIntegerOption(rawOption)) {
    return rawIntegerOptionToOption(rawOption);
  }
  if (isRawMultipleChoiceOption(rawOption)) {
    return rawMultipleChoiceOptionToOption(rawOption);
  }
  if (isRawChoiceOption(rawOption)) {
    return rawChoiceOptionToOption(rawOption);
  }
  if (isRawDateOption(rawOption)) {
    return rawDateOptionToOption(rawOption);
  }
  return rawBasicOptionToOption(rawOption);
};

const keysToSkip: string[] = ['service_type', 'page'];

export const normalize = (response: OptionsResponse): Options => {
  const rawOptions = response.actions.GET;
  return Object.keys(rawOptions).reduce((options, key) => {
    if (keysToSkip.includes(key)) {
      return options;
    }
    const rawOption = rawOptions[key];
    const option = rawOptionToOption(rawOption);
    return { ...options, [key]: option };
  }, {});
};
