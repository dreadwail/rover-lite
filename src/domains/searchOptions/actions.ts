import { Action } from 'redux';

import { request } from '../../api';
import { Options } from '../../state/searchOptions/types';
import { SearchType } from '../../state/searchTypes/types';
import { getEndpointUrl } from '../searchTypes/selectors';
import { PromiseThunkAction } from '../types';

import { normalize, OptionsResponse } from './normalize';

export interface FetchSearchOptionsPendingAction extends Action {
  readonly type: 'FETCH_SEARCH_OPTIONS.PENDING';
  readonly payload: {
    readonly searchType: SearchType;
  };
}

export interface FetchSearchOptionsFulfilledAction extends Action {
  readonly type: 'FETCH_SEARCH_OPTIONS.FULFILLED';
  readonly payload: {
    readonly searchType: SearchType;
    readonly options: Options;
  };
}

export interface FetchSearchOptionsRejectedAction extends Action {
  readonly type: 'FETCH_SEARCH_OPTIONS.REJECTED';
  readonly payload: {
    readonly searchType: SearchType;
    readonly error: any;
  };
  readonly error: true;
}

const fetchSearchOptionsPending = (searchType: SearchType): FetchSearchOptionsPendingAction => ({
  type: 'FETCH_SEARCH_OPTIONS.PENDING',
  payload: {
    searchType,
  },
});

const fetchSearchOptionsFulfilled = (
  searchType: SearchType,
  options: Options
): FetchSearchOptionsFulfilledAction => ({
  type: 'FETCH_SEARCH_OPTIONS.FULFILLED',
  payload: {
    searchType,
    options,
  },
});

const fetchSearchOptionsRejected = (
  searchType: SearchType,
  error: any
): FetchSearchOptionsRejectedAction => ({
  type: 'FETCH_SEARCH_OPTIONS.REJECTED',
  payload: {
    searchType,
    error,
  },
  error: true,
});

export type FetchSearchOptions = (searchType: SearchType) => PromiseThunkAction<void>;
export const fetchSearchOptions: FetchSearchOptions = searchType => async (dispatch, getState) => {
  dispatch(fetchSearchOptionsPending(searchType));
  try {
    const state = getState();
    const endpointUrl = getEndpointUrl(state, searchType);
    if (!endpointUrl) {
      throw new Error(`Unhandled search type: ${searchType}`);
    }
    const response = await request<OptionsResponse>('OPTIONS', endpointUrl);
    const normalizedOptions = normalize(response);
    dispatch(fetchSearchOptionsFulfilled(searchType, normalizedOptions));
  } catch (err) {
    dispatch(fetchSearchOptionsRejected(searchType, err));
  }
};
