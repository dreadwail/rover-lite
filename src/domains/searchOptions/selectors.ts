import { Options, OptionsByCategoryId, OptionCategory } from '../../state/searchOptions/types';
import { RoverSearchState } from '../../state/types';

export const hasFailed = (state: RoverSearchState): boolean => state.searchOptions.hasFailed;
export const isFetching = (state: RoverSearchState): boolean => state.searchOptions.isFetching;

const getOptions = (state: RoverSearchState): Options => state.searchOptions.options;

export const getRankedCategories = (state: RoverSearchState): OptionCategory[] => {
  const options = getOptions(state);
  const categoriesByKey = Object.keys(options).reduce<Record<string, OptionCategory>>(
    (categories, key) => {
      const { category } = options[key];
      return { ...categories, [category.categoryId]: category };
    },
    {}
  );
  return Object.values(categoriesByKey).sort((a, b) => a.rank - b.rank);
};

export const getOptionsByCategoryId = (state: RoverSearchState): OptionsByCategoryId => {
  const options = getOptions(state);
  return Object.keys(options).reduce<OptionsByCategoryId>((byCategoryId, key) => {
    const option = options[key];
    const { category } = option;
    const optionsHere = byCategoryId[category.categoryId];
    const newOptionsHere = { ...optionsHere, [key]: option };
    return {
      ...byCategoryId,
      [category.categoryId]: newOptionsHere,
    };
  }, {});
};
