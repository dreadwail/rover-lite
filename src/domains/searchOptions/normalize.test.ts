import { Options } from '../../state/searchOptions/types';

import { normalize } from './normalize';
import overnightBoarding from './__fixtures__/overnight-boarding.json';

describe('normalize', () => {
  let normalized: Options;

  describe('with a real overnight boarding payload', () => {
    beforeEach(() => {
      normalized = normalize(overnightBoarding);
    });

    it('returns the expected normalized values', () => {
      expect(normalized).toMatchSnapshot();
    });
  });
});
