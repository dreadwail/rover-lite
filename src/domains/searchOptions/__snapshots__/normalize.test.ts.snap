// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalize with a real overnight boarding payload returns the expected normalized values 1`] = `
Object {
  "accepts_only_one_client": Object {
    "category": Object {
      "categoryId": "pets_in_the_home",
      "displayName": "Pets in the home",
      "rank": 5,
    },
    "helpText": undefined,
    "label": "Accepts only one client at a time",
    "rank": 0,
    "required": false,
    "type": "boolean",
  },
  "apse": Object {
    "category": Object {
      "categoryId": "organizations",
      "displayName": "Organizations",
      "rank": 8,
    },
    "helpText": undefined,
    "label": "APSE member",
    "rank": 3,
    "required": false,
    "type": "boolean",
  },
  "bathing_grooming": Object {
    "category": Object {
      "categoryId": "advanced_services",
      "displayName": "Services",
      "rank": 9,
    },
    "helpText": undefined,
    "label": "Bathing / Grooming",
    "rank": 5,
    "required": false,
    "type": "boolean",
  },
  "cat_care": Object {
    "category": Object {
      "categoryId": "advanced_services",
      "displayName": "Services",
      "rank": 9,
    },
    "helpText": undefined,
    "label": "Cat care",
    "rank": 1,
    "required": false,
    "type": "boolean",
  },
  "centerlat": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 0,
    },
    "helpText": undefined,
    "label": "Centerlat",
    "rank": 0,
    "required": false,
    "type": "decimal",
  },
  "centerlng": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 0,
    },
    "helpText": undefined,
    "label": "Centerlng",
    "rank": 0,
    "required": false,
    "type": "decimal",
  },
  "dog_preference": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 0,
    },
    "choices": Array [
      Object {
        "displayName": "No other dogs",
        "value": "no_other_dogs",
      },
      Object {
        "displayName": "Other dogs",
        "value": "other_dogs",
      },
      Object {
        "displayName": "No preference",
        "value": "no_pref",
      },
    ],
    "helpText": undefined,
    "label": "Dog preference",
    "rank": 0,
    "required": false,
    "type": "choice",
  },
  "dogs_allowed_on_bed": Object {
    "category": Object {
      "categoryId": "housing_conditions",
      "displayName": "Housing conditions",
      "rank": 7,
    },
    "helpText": undefined,
    "label": "Dogs allowed on bed",
    "rank": 4,
    "required": false,
    "type": "boolean",
  },
  "dogs_allowed_on_furniture": Object {
    "category": Object {
      "categoryId": "housing_conditions",
      "displayName": "Housing conditions",
      "rank": 7,
    },
    "helpText": undefined,
    "label": "Dogs allowed on furniture",
    "rank": 5,
    "required": false,
    "type": "boolean",
  },
  "end_date": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 1,
    },
    "helpText": undefined,
    "label": "End date",
    "rank": 2,
    "required": false,
    "type": "date",
  },
  "fulltime_availability": Object {
    "category": Object {
      "categoryId": "daytime_availability",
      "displayName": "Daytime Availability",
      "rank": 2,
    },
    "helpText": undefined,
    "label": "Sitter is home full-time",
    "rank": 1,
    "required": false,
    "type": "boolean",
  },
  "has_fenced_yard": Object {
    "category": Object {
      "categoryId": "housing_conditions",
      "displayName": "Housing conditions",
      "rank": 7,
    },
    "helpText": undefined,
    "label": "Has fenced yard",
    "rank": 2,
    "required": false,
    "type": "boolean",
  },
  "has_house": Object {
    "category": Object {
      "categoryId": "housing_conditions",
      "displayName": "Housing conditions",
      "rank": 7,
    },
    "helpText": undefined,
    "label": "Has house (excludes apartments)",
    "rank": 1,
    "required": false,
    "type": "boolean",
  },
  "has_no_children": Object {
    "category": Object {
      "categoryId": "advanced_sitter_info",
      "displayName": "Sitter Info",
      "rank": 4,
    },
    "helpText": undefined,
    "label": "Has no children",
    "rank": 0,
    "required": false,
    "type": "boolean",
  },
  "is_premier": Object {
    "category": Object {
      "categoryId": "advanced_services",
      "displayName": "Services",
      "rank": 9,
    },
    "helpText": undefined,
    "label": "RoverGO",
    "rank": 4,
    "required": false,
    "type": "boolean",
  },
  "knows_first_aid": Object {
    "category": Object {
      "categoryId": "housing_conditions",
      "displayName": "Housing conditions",
      "rank": 7,
    },
    "helpText": undefined,
    "label": "Dog first-aid certified",
    "rank": 3,
    "required": false,
    "type": "boolean",
  },
  "maxlat": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 0,
    },
    "helpText": undefined,
    "label": "Maxlat",
    "rank": 0,
    "required": false,
    "type": "decimal",
  },
  "maxlng": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 0,
    },
    "helpText": undefined,
    "label": "Maxlng",
    "rank": 0,
    "required": false,
    "type": "decimal",
  },
  "maxprice": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 1,
    },
    "helpText": undefined,
    "label": "Maximum Price",
    "maxValue": undefined,
    "minValue": undefined,
    "rank": 5,
    "required": false,
    "type": "integer",
  },
  "minlat": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 0,
    },
    "helpText": undefined,
    "label": "Minlat",
    "rank": 0,
    "required": false,
    "type": "decimal",
  },
  "minlng": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 0,
    },
    "helpText": undefined,
    "label": "Minlng",
    "rank": 0,
    "required": false,
    "type": "decimal",
  },
  "minprice": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 1,
    },
    "helpText": undefined,
    "label": "Minimum Price",
    "maxValue": undefined,
    "minValue": undefined,
    "rank": 4,
    "required": false,
    "type": "integer",
  },
  "no_caged_pets": Object {
    "category": Object {
      "categoryId": "pets_in_the_home",
      "displayName": "Pets in the home",
      "rank": 5,
    },
    "helpText": undefined,
    "label": "Does not own caged pets",
    "rank": 3,
    "required": false,
    "type": "boolean",
  },
  "no_cats": Object {
    "category": Object {
      "categoryId": "pets_in_the_home",
      "displayName": "Pets in the home",
      "rank": 5,
    },
    "helpText": undefined,
    "label": "Doesn't own a cat",
    "rank": 2,
    "required": false,
    "type": "boolean",
  },
  "no_children_0_5": Object {
    "category": Object {
      "categoryId": "children_in_the_home",
      "displayName": "Children in the home",
      "rank": 6,
    },
    "helpText": undefined,
    "label": "No children 0-5 years old",
    "rank": 1,
    "required": false,
    "type": "boolean",
  },
  "no_children_6_12": Object {
    "category": Object {
      "categoryId": "children_in_the_home",
      "displayName": "Children in the home",
      "rank": 6,
    },
    "helpText": undefined,
    "label": "No children 6-12 years old",
    "rank": 2,
    "required": false,
    "type": "boolean",
  },
  "non_smoking": Object {
    "category": Object {
      "categoryId": "housing_conditions",
      "displayName": "Housing conditions",
      "rank": 7,
    },
    "helpText": undefined,
    "label": "Non-smoking home",
    "rank": 6,
    "required": false,
    "type": "boolean",
  },
  "person_does_not_have_dogs": Object {
    "category": Object {
      "categoryId": "pets_in_the_home",
      "displayName": "Pets in the home",
      "rank": 5,
    },
    "helpText": undefined,
    "label": "Doesn't own a dog",
    "rank": 1,
    "required": false,
    "type": "boolean",
  },
  "pet_sizes": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 0,
    },
    "choices": Array [
      Object {
        "displayName": "0-15",
        "value": "small",
      },
      Object {
        "displayName": "16-40",
        "value": "medium",
      },
      Object {
        "displayName": "41-100",
        "value": "large",
      },
      Object {
        "displayName": "101+",
        "value": "giant",
      },
    ],
    "helpText": undefined,
    "label": "How large is your dog (in lbs)?",
    "rank": 0,
    "required": false,
    "type": "multiple choice",
  },
  "petsitusa": Object {
    "category": Object {
      "categoryId": "organizations",
      "displayName": "Organizations",
      "rank": 8,
    },
    "helpText": undefined,
    "label": "PetsitUSA member",
    "rank": 2,
    "required": false,
    "type": "boolean",
  },
  "puppy": Object {
    "category": Object {
      "categoryId": "advanced_services",
      "displayName": "Services",
      "rank": 9,
    },
    "helpText": undefined,
    "label": "Puppy care",
    "rank": 2,
    "required": false,
    "type": "boolean",
  },
  "residence_preference": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 0,
    },
    "choices": Array [
      Object {
        "displayName": "Houses",
        "value": "houses",
      },
      Object {
        "displayName": "Apartments",
        "value": "apartments",
      },
      Object {
        "displayName": "No preference",
        "value": "no_pref",
      },
    ],
    "helpText": undefined,
    "label": "Residence preference",
    "rank": 0,
    "required": false,
    "type": "choice",
  },
  "senior_dog_care": Object {
    "category": Object {
      "categoryId": "advanced_services",
      "displayName": "Services",
      "rank": 9,
    },
    "helpText": undefined,
    "label": "Senior dog care",
    "rank": 3,
    "required": false,
    "type": "boolean",
  },
  "start_date": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 1,
    },
    "helpText": undefined,
    "label": "Start date",
    "rank": 1,
    "required": false,
    "type": "date",
  },
  "test_key_no_category_display_name": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 0,
    },
    "helpText": undefined,
    "label": "Test No Category Display Name",
    "rank": 7,
    "required": false,
    "type": "boolean",
  },
  "test_key_no_category_rank": Object {
    "category": Object {
      "categoryId": "detail",
      "displayName": "Detail",
      "rank": 0,
    },
    "helpText": undefined,
    "label": "Test No Category Rank",
    "rank": 6,
    "required": false,
    "type": "boolean",
  },
  "volunteer_donor": Object {
    "category": Object {
      "categoryId": "organizations",
      "displayName": "Organizations",
      "rank": 8,
    },
    "helpText": undefined,
    "label": "Volunteer / Donor",
    "rank": 1,
    "required": false,
    "type": "boolean",
  },
}
`;
